File: solutions/Sliding Window/239_sliding-window-maximum.py
from typing import List
import heapq
from collections import deque

class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        res = []
        max_heap = []
        deleted = {}

        for i in range(k):
            heapq.heappush(max_heap, -nums[i])

        res.append(-max_heap[0])
        l = 0
        for r in range(k, len(nums)):
            heapq.heappush(max_heap, -nums[r])

            deleted[nums[l]] = deleted.get(nums[l], 0) + 1
            l += 1

            while deleted.get(-max_heap[0], 0):
                deleted[-max_heap[0]] -= 1
                heapq.heappop(max_heap)

            res.append(-max_heap[0])

        return res
    
# Time Complexity: O(n log k)
# Space Complexity: O(k)


class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        queue = deque()
        res = []

        for i in range(k):
            while queue and queue[-1] < nums[i]:
                queue.pop()

            queue.append(nums[i])

        res.append(queue[0])
        for r in range(k, len(nums)):
            while queue and queue[-1] < nums[r]:
                queue.pop()

            queue.append(nums[r])
            if queue[0] == nums[r - k]:
                queue.popleft()

            res.append(queue[0])

        return res
    
# Time Complexity: O(n)
# Space Complexity: O(k)

File: C++/Sliding Window/239_sliding-window-maximum.cpp
#include <deque>
#include <vector>
#include <unordered_set>
#include <queue>

using namespace std;

class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        int n = nums.size();
        deque<int> window;
        for (int i = 0; i < k - 1; i++) {
            while (!window.empty() && window.back() < nums[i]) window.pop_back();
            window.push_back(nums[i]);
        }

        vector<int> res;
        for (int i = k - 1; i < n; i++) {
            while (!window.empty() && window.back() < nums[i]) window.pop_back();
            window.push_back(nums[i]);

            res.push_back(window.front());
            if (window.front() == nums[i - k + 1]) window.pop_front();
        }

        return res;
    }
};


class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        int n = nums.size();
        unordered_set<int> deleted;
        priority_queue<pair<int, int>, vector<pair<int, int>>, less<pair<int, int>>> heap;
        for (int i = 0; i < k - 1; i++) {
            heap.push({nums[i], i});
        }

        vector<int> res;
        for (int i = k - 1; i < n; i++) {
            heap.push({nums[i], i});

            while (deleted.contains(heap.top().second)) heap.pop();

            res.push_back(heap.top().first);
            deleted.insert(i - k + 1);
        }

        return res;
    }
};